index/file_category.go

```diff
@@ -0,0 +1,97 @@
+package index
+
+import (
+	"errors"
+
+	"github.com/go-enry/go-enry/v2"
+)
+
+// FileCategory represents the category of a file, as determined by go-enry. It is non-exhaustive
+// but tries to the major cases like whether the file is a test, generated, etc.
+//
+// A file's category is used in search scoring to determine the weight of a file match.
+type FileCategory byte
+
+const (
+	FileCategoryMissing FileCategory = iota
+	FileCategoryDefault
+	FileCategoryTest
+	FileCategoryVendored
+	FileCategoryGenerated
+	FileCategoryConfig
+	FileCategoryDotFile
+	FileCategoryDocumentation
+	// FileCategoryMissing is a sentinel value that indicates we never computed the file category during indexing
+	// (which means we're reading from an old index version). This value can never be written to the index.
+)
+
+func DetermineFileCategory(name string, content []byte) FileCategory {
+	// Note: the order of these checks is important.
+	// - IsDotFile must come before IsVendor
+	if enry.IsTest(name) {
+		return FileCategoryTest
+	} else if enry.IsDotFile(name) {
+		return FileCategoryDotFile
+	} else if enry.IsVendor(name) {
+		return FileCategoryVendored
+	} else if enry.IsGenerated(name, content) {
+		return FileCategoryGenerated
+	} else if enry.IsConfiguration(name) {
+		return FileCategoryConfig
+	} else if enry.IsDotFile(name) {
+		return FileCategoryDotFile
+	} else if enry.IsDocumentation(name) {
+		return FileCategoryDocumentation
+	}
+	return FileCategoryDefault
+}
+
+// lowPriority returns true if this file category is considered 'low priority'. This is used
+// in search scoring to down-weight matches in these files.
+func (c FileCategory) lowPriority() bool {
+	return c == FileCategoryTest || c == FileCategoryVendored || c == FileCategoryGenerated
+}
+
+func (c FileCategory) encode() (byte, error) {
```

Comment:
This is a bit overkill, since `FileCategory` is a byte and we could just cast it. However, I like being really explicit like this. Otherwise, someone could accidentally reorder the enum values, and then the index categories would be wrong.